### PR TITLE
Operator Workflows and Tests

### PR DESCRIPTION
**Pull request by phudtran**
This PR adds the following:

- Workflows for interface between Operator and Daemon.
- Workflows for Scale up/down of FWDs, Chains, and FTNs
- Functional tests for operator
- Rework of Chain RPC and unit tests to include tail FTN IP and MAC
- Add control IP to droplet CRD for RPC with Daemon

### CODE REVIEW

File: deploy/etc/crds/dfts.crd.yaml [hunk 1]
```diff
@@ -28,6 +28,14 @@ spec:
           type: array
           items:
             type: string
+        chains:
+          type: array
+          items:
+            type: string
+        fwds:
```
Comment (liangbin-pub) [inline deploy/etc/crds/dfts.crd.yaml:35]: Should we associate fwd to dft or dft to fwd?

File: deploy/etc/crds/droplets.crd.yaml [hunk 2]
```diff
@@ -49,6 +49,11 @@ spec:
       priority: 0
       JSONPath: .spec.itf
       description: The main interface of the droplet
+    - name: IpControl
+      type: string
+      priority: 0
+      JSONPath: .spec.ip_control
+      description: The Node ZGC interface ip
```
Comment (liangbin-pub) [inline deploy/etc/crds/droplets.crd.yaml:56]: The ZGC node interface ip
BTW, do we need the "Interface" info above?

File: deploy/kind/create_zgc.sh [hunk 3]
```diff
@@ -54,22 +52,22 @@ manager_ip=$(kubectl get node $(kubectl get pods -o wide | grep zeta-manager | a
...
 response=$(curl -H 'Content-Type: application/json' -X POST \
     -d '{"name":"zgc0",
-        "description":"zgc0",
-        "ip_start":"20.0.0.1",
-        "ip_end":"20.0.0.16",
-        "port_ibo":"8300"}' \
+            "description":"zgc0",
+            "ip_start":"20.0.0.0",
+            "ip_end":"20.0.0.255",
+            "port_ibo":"8300"}' \
```
Comment (liangbin-pub) [inline deploy/kind/create_zgc.sh]: unnecessary large range but no harm either

File: deploy/kind/create_zgc.sh [hunk 4]
```diff
@@ -54,22 +52,22 @@ manager_ip=$(kubectl get node $(kubectl get pods -o wide | grep zeta-manager | a
...
     $manager_ip:80/zgcs)
 
 zgc_id=$(echo $response | sed 's/\\[tn]//g' | cut -d':' -f 7 | tr -d '"}' | xargs)
-echo "REST API: Test ZGC zgc0 created with UUID: $zgc_id returned"
+tenant_network="tenant_network"
+zgc_network="zgc_network"
 
-# Register ZGC nodes containers in KIND deployment
+# Register node containers in KIND deployment
 inf_control="eth0"
```
Comment (liangbin-pub) [inline deploy/kind/create_zgc.sh:67]: Original is more accurate to distinguish between compute nodes and ZGC nodes

File: deploy/kind/create_zgc.sh [hunk 5]
```diff
@@ -54,22 +52,22 @@ manager_ip=$(kubectl get node $(kubectl get pods -o wide | grep zeta-manager | a
...
+            "description":"zgc0",
+            "ip_start":"20.0.0.0",
+            "ip_end":"20.0.0.255",
+            "port_ibo":"8300"}' \
     $manager_ip:80/zgcs)
 
 zgc_id=$(echo $response | sed 's/\\[tn]//g' | cut -d':' -f 7 | tr -d '"}' | xargs)
-echo "REST API: Test ZGC zgc0 created with UUID: $zgc_id returned"
+tenant_network="tenant_network"
+zgc_network="zgc_network"
```
Comment (liangbin-pub) [inline deploy/kind/create_zgc.sh:65]: these are already defined at line 10,11 and unnecessary change for echo for indication of ZGC POST result

File: deploy/kind/create_zgc.sh [hunk 6]
```diff
@@ -54,22 +52,23 @@ manager_ip=$(kubectl get node $(kubectl get pods -o wide | grep zeta-manager | a
 
 response=$(curl -H 'Content-Type: application/json' -X POST \
     -d '{"name":"zgc0",
-        "description":"zgc0",
-        "ip_start":"20.0.0.1",
-        "ip_end":"20.0.0.16",
-        "port_ibo":"8300"}' \
+            "description":"zgc0",
+            "ip_start":"20.0.0.0",
```
Comment (liangbin-pub) [inline deploy/kind/create_zgc.sh:56]: avoid use .0 for unicast IP, 1-16 are good

File: deploy/kind/create_zgc.sh [hunk 7]
```diff
@@ -78,17 +77,17 @@ do
...
+        "description":"'"$node"'",
+        "name":"'"$node"'",
+        "ip_control":"'"$ip_control"'",
+        "id_control":"ubuntu",
+        "pwd_control":"changeme",
+        "inf_tenant":"'"$inf_tenant"'",
+        "mac_tenant":"'"$mac_tenant"'",
+        "inf_zgc":"'"$inf_zgc"'",
+        "mac_zgc":"'"$mac_zgc"'"}'
     curl -H 'Content-Type: application/json' -X POST -d "$body" $manager_ip:80/nodes
```
Comment (liangbin-pub) [inline deploy/kind/create_zgc.sh:89]: not clear what changed

File: deploy/playbooks/inventories/vars/site_kind.yml [hunk 8]
```diff
@@ -9,4 +9,4 @@
 #
 k8s_type: "kind"
 k8s_size: 2
-droplet_size: 3
+droplet_size: 1
```
Comment (liangbin-pub) [inline deploy/playbooks/inventories/vars/site_kind.yml:12]: so we don't want to have 3 nodes ZGC or is this change just for debugging?

File: src/cli/trn_cli_chain.c [hunk 9]
```diff
@@ -31,6 +33,30 @@ int trn_cli_parse_chain(const cJSON *jsonobj, struct rpc_trn_chain_t *chain)
...
+		return -EINVAL;
+	}
+	if (tail_ftn_mac != NULL && cJSON_IsString(tail_ftn_mac)) {
+		if (6 ==
+		    sscanf(tail_ftn_mac->valuestring,
+			   "%hhx:%hhx:%hhx:%hhx:%hhx:%hhx%*c",
+			   &chain->tail_ftn_mac[0], &chain->tail_ftn_mac[1],
+			   &chain->tail_ftn_mac[2], &chain->tail_ftn_mac[3],
+			   &chain->tail_ftn_mac[4], &chain->tail_ftn_mac[5])) {
+		} else {
```
Comment (liangbin-pub) [inline src/cli/trn_cli_chain.c:51]: there are API added in trn_cli_common.c to parse IP, MAC strings

File: src/mgmt/manager/entrypoint.sh [hunk 10]
```diff
@@ -19,12 +19,4 @@ echo "PostgreSQL started"
...
-mkdir -p /var/log/gunicorn/
-
-touch /var/log/gunicorn/error.log
-
-touch /var/log/gunicorn/access.log
-
-
-gunicorn -b 0.0.0.0:5000 manage:app --error-logfile /var/log/gunicorn/error.log --access-logfile /var/log/gunicorn/access.log --capture-output --log-level debug
-
+gunicorn -b 0.0.0.0:5000 manage:app --error-logfile - --access-logfile - --log-level info
```
Comment (liangbin-pub) [inline src/mgmt/manager/entrypoint.sh:22]: Will this making python stdio/stderr goes to kubectl logs? somehow after revious log change we lost that capability. 
For log level, should we keep debug or better based on image build or deployment variables?

File: src/mgmt/manager/project/__init__.py [hunk 11]
```diff
@@ -29,10 +30,6 @@ def create_app(script_info=None):
 
     # instantiate the app
     app = Flask(__name__)
-    gunicorn_error_logger = logging.getLogger('gunicorn.error')
-    app.logger.handlers.extend(gunicorn_error_logger.handlers)
-    app.logger.setLevel(logging.DEBUG)
-    app.logger.debug('this will show in the log')
     # enable CORS
```
Comment (liangbin-pub) [inline src/mgmt/manager/project/__init__.py:33]: This log level (python) is different from  level set in gunicorn command above.

File: src/mgmt/manager/project/api/nodes.py [hunk 12]
```diff
@@ -36,27 +36,30 @@ def update_droplet(droplet):
...
                                                                    name=droplet['metadata']['name'],
                                                                    body=droplet)
         logger.info('Response for update droplet: {}'.format(update_response))
-        droplet_update_itf_config(droplet['metadata']['labels']['node_ip'], droplet)
+        droplet_update_itf_config(
+            droplet['metadata']['labels']['node_ip'], droplet)
     except ApiException as e:
         logger.error('Exception when updating existing droplet: {}'.format(e))
 
+
```
Comment (liangbin-pub) [inline src/mgmt/manager/project/api/nodes.py:48]: why do we add so many whitespace changes?

File: src/mgmt/manager/project/api/nodes.py [hunk 13]
```diff
@@ -279,10 +298,11 @@ def single_node(node_id):
                                                                  version='v1',
                                                                  plural='droplets',
                                                                  label_selector='zgc_id='+zgc_id+',network=tenant'
-                                                                )['items']
+                                                                 )['items']
```
Comment (liangbin-pub) [inline src/mgmt/manager/project/api/nodes.py:301]: is all these whitespace changes generated by some lint tool? the right parenthesis matches with left and adding a space makes not align. 
I really think we should separate the changes so we can have a dedicated commit just to fix all "non-standard" whitespaces cross codebase.  It's hard to tell real functional changes from these formatting changes (which a dedicate PR will requires almost no effort to review).

File: src/mgmt/manager/project/api/ports.py [hunk 14]
```diff
@@ -30,8 +31,10 @@
 
 eps = []
 
+
 def ports_update_eps_config():
-    eps_list = [eps[i:i + EP_BATCH_MAX] for i in range(0, len(eps), EP_BATCH_MAX)]
+    eps_list = [eps[i:i + EP_BATCH_MAX]
+                for i in range(0, len(eps), EP_BATCH_MAX)]
     for eps_chunk in eps_list:
```
Comment (liangbin-pub) [inline src/mgmt/manager/project/api/ports.py:38]: I have to say it as in current days, 80 columns shouldn't be a hard restriction any more, a 100 or more often 120 wide is often used to maintain code continuity and readability. Suggest we adopt the 120 rule instead of the 80 rule for old punchcard days.

File: src/mgmt/manager/project/api/ports.py [hunk 15]
```diff
@@ -53,16 +57,17 @@ def all_ports():
...
+                ip_node=post_data.get('ip_node')).first()
             if host is None:
                 host = Host(
-                    mac_node = post_data.get('mac_node'),
-                    ip_node = post_data.get('ip_node'))
-                host.host_id=str(uuid.uuid4())
+                    mac_node=post_data.get('mac_node'),
+                    ip_node=post_data.get('ip_node'))
+                host.host_id = str(uuid.uuid4())
                 db.session.add(host)
```
Comment (liangbin-pub) [inline src/mgmt/manager/project/api/ports.py:71]: Again, for these kind of whitespace changes, sometimes spaces are added around operator, sometimes are removed

File: src/mgmt/manager/project/api/settings.py [hunk 16]
```diff
@@ -23,12 +24,16 @@ def init():
...
+        )
         logger.info('Result for getting configmap: {}'.format(api_response))
-        zgc_cidr_range = api_response.items[0].data['cidr']
-        MAC_PRIVATE = int(api_response.items[0].data['MAC_PRIVATE'],0)
+        zgc_cidr_range = api_response.data['cidr']
+        MAC_PRIVATE = int(api_response.data['MAC_PRIVATE'], 0)
     except ApiException as e:
-        logger.error('Exception when trying to get zgc-cluster-config: {}'.format(e))
+        logger.error(
+            'Exception when reading configmap {} : {}'.format("zgc-cluster-config", e))
```
Comment (liangbin-pub) [inline src/mgmt/manager/project/api/settings.py:39]: why passing a const string for formatter

File: src/mgmt/operator/common/id_allocator.py [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+        for i in range(1, 10000):  # id 0 reserved for DFT
+            self.ids.append(str(i))
+
+    def allocate_id(self, obj_type):
+        logger.info("Allocating ID {} for {}!".format(self.ids[0], obj_type))
+        return (self.ids.pop(0))
+
+    def reclaim_id(self, obj_id, obj_type):
+        logger.info("Reclaiming ID {} from {}!".format(obj_id, obj_type))
+        self.ids.insert(0, obj_id)
```
Comment (liangbin-pub) [inline src/mgmt/operator/common/id_allocator.py:33]: Suggest put this to common module for manager as well

File: src/mgmt/operator/workflows/dfts/create.py [hunk 18]
```diff
@@ -5,14 +5,29 @@
...
+        task.raise_temporary_error(
+            "No droplets available for DFT")
+    if dft.numchains < 1:
+        task.raise_permanent_error("DFT numchains cannot be less than 1")
+    chains_opr.create_n_chains(
+        dft, dft.numchains, OBJ_DEFAULTS.default_n_ftns, task)
+    if dft.numfwds < 1:
+        task.raise_permanent_error("DFT numchains cannot be less than 1")
+    fwds_opr.create_n_fwds(
+        dft, dft.numfwds, task)
```
Comment (liangbin-pub) [inline src/mgmt/operator/workflows/dfts/create.py:31]: Single line is much readable. Also, shall we bail out for above errors earlier?

File: src/mgmt/operator/workflows/ftns/delete.py [hunk 19]
```diff
@@ -5,10 +5,19 @@
...
-def ftn_delete(task, ftn, name, body, spec):
+def ftn_delete(task, ftn, name, body, spec, diff):
     logger.info("Deleting Ftn {}!".format(name))
+    if not ftn:
+        ftn = ftns_opr.get_stored_obj(name, spec)
+    droplet_obj = droplets_opr.store.get_obj(ftn.droplet, KIND.droplet)
+
+    # Don't need to run delete_ftn, unload of XDP cleans it up.
+    droplet_obj.rpc.unload_transit_xdp()
+    droplets_opr.unassign_droplet(ftn)
```
Comment (liangbin-pub) [inline src/mgmt/operator/workflows/ftns/delete.py:22]: Suggest adjust current xdp load/unload hook to make this work

File: src/mgmt/operator/workflows/fwds/create.py [hunk 20]
```diff
@@ -8,21 +8,18 @@
...
     if not fwd:
         fwd = fwds_opr.get_stored_obj(name, spec)
     if not dfts_opr.store.contains_obj(fwd.dft, KIND.dft):
         task.raise_temporary_error(
             "DFT {} not yet created!".format(fwd.dft))
-    if len(droplets_opr.get_unallocated_droplets()) < 1:
-        task.raise_temporary_error(
-            "No droplets available for FWD")
-    droplets_opr.assign_droplet(fwd)
     return fwd
```
Comment (liangbin-pub) [inline src/mgmt/operator/workflows/fwds/create.py:25]: Should we move transit XDP load rpc to here?

File: src/xdp/trn_transit_xdp.c [hunk 21]
```diff
@@ -513,11 +544,12 @@ int _transit(struct xdp_md *ctx)
...
 	pkt.itf_idx = ctx->ingress_ifindex;
-	
+
 	pkt.itf = bpf_map_lookup_elem(&if_config_map, &pkt.itf_idx);
 	if (!pkt.itf) {
-		bpf_debug("[Transit:%d] ABORTED: Failed to lookup ingress config for %d\n",
-			  __LINE__, pkt.itf_idx);
+		bpf_debug(
+			"[Transit:%d] ABORTED: Failed to lookup ingress config for %d\n",
+			__LINE__, pkt.itf_idx);
```
Comment (liangbin-pub) [inline src/xdp/trn_transit_xdp.c:552]: I tried to figure out exactly what's the change in this file line by line but looks like all whitespace changes, maybe I missed something but I suggest we split the change set into two, one for functional/bug change and one for just format change then we use 2nd change set to fix all whitespaces in repo against some lint files. This will make review/merge easier for the future.

**Review by liangbin-pub — COMMENTED** [review]
Overwhelmed by the whitespace changes, may missed some actual functional changes. Old days we follow a rule called "If It Ain't Broke, Don't Fix It" to minimize pure whitespace changes which makes merge/review much harder, this is especially true to avoid unintentional changes caused by some editor automatically, which behaves differently on each setup. 
I do appreciate it to make the codes following some consistent standard, but we can do it in a cleaner way by enforcing same dev environment with same rules for new changes, maybe automatically. But for existing code, I suggest not  mixing with actual functional change and do it separately as a catch-up-to-standard effort.
Regarding changes in operator, my suggestion is to move the current XDP load/unload from manager API handler to operator workflow.
Another thing may need attention is that droplet != FWD or FTN. XDP on droplet includes many functionalities and FWD/FTN are just part of those.
What do you think?

File: src/mgmt/manager/entrypoint.sh (hunk 10 above)
Comment (zzxgzgz) [reply src/mgmt/manager/entrypoint.sh:22]: The log part looks good to me.

File: deploy/etc/crds/dfts.crd.yaml (hunk 1 above)
Comment (phudtran) [reply deploy/etc/crds/dfts.crd.yaml:35]: Right now DFTs have a list of all available FWDs, and all FWD's are associated with a single DFT. In the future we can expand this to have FWDs associated with a list of DFTs if needed.

File: deploy/etc/crds/droplets.crd.yaml (hunk 2 above)
Comment (phudtran) [reply deploy/etc/crds/droplets.crd.yaml:56]: Thanks for the catch. Yes we do this need this in order to make the RPC calls as the operator running in the Kubernetes is not on the same network as the droplet.

File: deploy/kind/create_zgc.sh (hunk 3 above)
Comment (phudtran) [reply deploy/kind/create_zgc.sh]: Oops, I thought I changed this. Meant to use 20.0.0.239 - 20.0.0.254 instead.

File: deploy/kind/create_zgc.sh (hunk 4 above)
Comment (phudtran) [reply deploy/kind/create_zgc.sh:67]: Will change back

File: deploy/kind/create_zgc.sh (hunk 5 above)
Comment (phudtran) [reply deploy/kind/create_zgc.sh:65]: Will change back.

File: deploy/kind/create_zgc.sh (hunk 6 above)
Comment (phudtran) [reply deploy/kind/create_zgc.sh:56]: Ok will do.

File: deploy/kind/create_zgc.sh (hunk 7 above)
Comment (phudtran) [reply deploy/kind/create_zgc.sh:89]: I think just formatting....

File: deploy/playbooks/inventories/vars/site_kind.yml (hunk 8 above)
Comment (phudtran) [reply deploy/playbooks/inventories/vars/site_kind.yml:12]: Yeah this was just for debugging.

File: src/cli/trn_cli_chain.c (hunk 9 above)
Comment (phudtran) [reply src/cli/trn_cli_chain.c:51]: Let me use those instead for consistency.

File: src/mgmt/manager/project/__init__.py (hunk 11 above)
Comment (phudtran) [reply src/mgmt/manager/project/__init__.py:33]: These lines were deleted.

File: src/mgmt/manager/project/api/nodes.py (hunk 12 above)
Comment (phudtran) [reply src/mgmt/manager/project/api/nodes.py:48]: Linter format on save :/

File: src/mgmt/manager/project/api/nodes.py (hunk 13 above)
Comment (phudtran) [reply src/mgmt/manager/project/api/nodes.py:301]: Agreed. Discussed offline

File: src/mgmt/manager/project/api/ports.py (hunk 14 above)
Comment (phudtran) [reply src/mgmt/manager/project/api/ports.py:38]: Agreed, I'm on 21:9 myself :)

File: src/mgmt/manager/project/api/settings.py (hunk 16 above)
Comment (phudtran) [reply src/mgmt/manager/project/api/settings.py:39]: Will fix, thanks for catch.

File: src/mgmt/operator/common/id_allocator.py (hunk 17 above)
Comment (phudtran) [reply src/mgmt/operator/common/id_allocator.py:33]: Sure I can move it there.

File: src/mgmt/operator/workflows/dfts/create.py (hunk 18 above)
Comment (phudtran) [reply src/mgmt/operator/workflows/dfts/create.py:31]: Agreed. I think we can bail out the "available droplets" error earlier. As for the other errors, we don't actually know them until the DFT is created.

File: src/mgmt/operator/workflows/ftns/delete.py (hunk 19 above)
Comment (phudtran) [reply src/mgmt/operator/workflows/ftns/delete.py:22]: Agreed. I changed my xdp load and unload after rebasing with your change since it was being done in the manager code.

File: src/mgmt/operator/workflows/fwds/create.py (hunk 20 above)
Comment (phudtran) [reply src/mgmt/operator/workflows/fwds/create.py:25]: Sure I can remove it from the manager code and move it here.

File: src/xdp/trn_transit_xdp.c (hunk 21 above)
Comment (phudtran) [reply src/xdp/trn_transit_xdp.c:552]: We will have a separate PR that will format all the code to an agreed upon standard by the team. Synced offline with the team.
